File: Tutorials/ObjC/Example11/Example11/ViewController.m
//
//  ViewController.m
//  Example11
//
//  Created by dev on 1/31/20.
//  Copyright © 2020 Penthera. All rights reserved.
//

#import "ViewController.h"

#import "DemoPlayerViewController.h"

// ---------------------------------------------------------------------------------------------------------
// IMPORTANT:
// The following three values must be initialzied, please contact support@penthera.com to obtain these keys
// ---------------------------------------------------------------------------------------------------------
static NSString* backplaneUrl = @"replace_with_your_backplane_url";
static NSString* publicKey = @"replace_with_your_public_key";
static NSString* privateKey = @"replace_with_your_private_key";

@interface ViewController ()

//
// MARK: Instance data
//
@property (nonatomic,strong) VirtuosoAsset *exampleAsset;
@property (nonatomic, strong) VirtuosoDownloadEngineNotificationManager* downloadEngineNotifications;
@property (nonatomic, strong) VirtuosoBackplaneNotificationsManager* backplaneNotifications;
@property (nonatomic, strong) NSError* error;
@property (nonatomic, assign) Boolean detectedDeviceUnregister;

//
// MARK: Outlets
//
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UIButton *downloadBtn;
@property (weak, nonatomic) IBOutlet UIButton *playBtn;
@property (weak, nonatomic) IBOutlet UIButton *deleteBtn;
@property (weak, nonatomic) IBOutlet UIProgressView *statusProgressBar;
@property (weak, nonatomic) IBOutlet UILabel *assetName;

@end

@implementation ViewController

//
// MARK: Lifecycle methods
//

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // download engine update listener
    self.downloadEngineNotifications = [[VirtuosoDownloadEngineNotificationManager alloc]initWithDelegate:self];
    self.backplaneNotifications = [[VirtuosoBackplaneNotificationsManager alloc]initWithDelegate:self];
    self.statusLabel.text = @"Starting Engine...";
    
    //
    // Enable the Engine
    VirtuosoDownloadEngine.instance.enabled = TRUE;
    
    // Backplane permissions require a unique user-id for the full range of captabilities support to work
    // Production code that needs this will need a unique customer ID.
    // For demonstation purposes only, we use the device name
    NSString* userName = UIDevice.currentDevice.name;
    
    //
    // Create the engine confuration
    VirtuosoEngineConfig* config = [[VirtuosoEngineConfig alloc]initWithUser:userName
                                                                backplaneUrl:backplaneUrl
                                                                   publicKey:publicKey
                                                                  privateKey:privateKey];
    if (!config)
    {
        dispatch_async(dispatch_get_main_queue(), ^{
            UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Setup Required"
                                                                           message:@"Please contact support@penthera.com to setup the backplaneUrl, pulicKey, and privateKey"
                                                                    preferredStyle:UIAlertControllerStyleAlert];
            UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                exit(0);
            }];
            [alert addAction:ok];
            [self presentViewController:alert animated:TRUE completion:nil];

        });
        return;
    }

    //
    // Start the Engine, this call only needs to happen one time.
    // This method will execute async, the callback will happen on the main-thread.
    [VirtuosoDownloadEngine.instance startup:config startupCallback:^(kVDE_EngineStartupCode status) {
        if (status != kVDE_EngineStartupSuccess)
        {
             UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Startup Error" message:[NSString stringWithFormat:@"Error starting the Engine. Status: %@", @(status)] preferredStyle:UIAlertControllerStyleAlert];
            [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            }]];
            [self presentViewController:alert animated:TRUE completion:nil];
        }
    }];

}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
}

// MARK: Click Handlers
- (IBAction)unregisterClicked:(id)sender {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        NSError* error = [self unregisterAllDevices];
        
        NSTimeInterval maximumWait = 30;
        NSTimeInterval sleepDuration = 1;
        while (!self.detectedDeviceUnregister && maximumWait > 0)
        {
            [NSThread sleepForTimeInterval:sleepDuration];
            maximumWait -= sleepDuration;
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            UIAlertController* alert = nil;
            if (error)
            {
                 alert = [UIAlertController alertControllerWithTitle:@"Unregister Error" message:[NSString stringWithFormat:@"Error while unregistering devices. Error: %@", error.localizedDescription] preferredStyle:UIAlertControllerStyleAlert];
                [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                }]];
            }
            else if (maximumWait <= 0)
            {
                alert = [UIAlertController alertControllerWithTitle:@"Unregister Complete With Issues" message:@"All devices are unregistered but never detected device unregister notification. App will now exit." preferredStyle:UIAlertControllerStyleAlert];
                [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                    exit(0);
                }]];
            }
            else
            {
                alert = [UIAlertController alertControllerWithTitle:@"Unregister Complete" message:@"All devices are unregistered. App will now exit." preferredStyle:UIAlertControllerStyleAlert];
                [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                    exit(0);
                }]];
            }
            [self presentViewController:alert animated:TRUE completion:nil];
        });
    });
}

-(NSError*)unregisterAllDevices
{
    VirtuosoDevice* currentDevice = nil;
    
    NSError* error = nil;
    for (VirtuosoDevice* device in VirtuosoDownloadEngine.instance.devices)
    {
        if (device.isThisDevice)
        {
            currentDevice = device;
        }
        else
        {
            error = [self unregisterDeviceAndWait:device];
            if (error)
            {
                return error;
            }
        }
    }
    
    if (currentDevice)
    {
        error = [self unregisterDeviceAndWait:currentDevice];
    }
    
    return error;
}

-(NSError*)unregisterDeviceAndWait:(VirtuosoDevice*)device
{
    __block NSError* result = nil;
    if (!device)
    {
        NSLog(@"Parameter 'device' is required");
        return result;
    }
    
    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        [device unregisterOnComplete:^(Boolean success, NSError * _Nullable error) {
            result = error;
            if (success)
            {
                NSLog(@"unregister succeeded for device: %@", device.nickname);
            }
            else
            {
                NSLog(@"unregister failed for device: %@ error: %@", device.nickname, error);
            }
            dispatch_semaphore_signal(semaphore);
        }];
    });
    
    const int waitTimeInSeconds = 30;
    if( dispatch_semaphore_wait(semaphore, dispatch_time(DISPATCH_TIME_NOW, (int64_t)waitTimeInSeconds * NSEC_PER_SEC)) != 0 )
    {
        NSLog(@"Timeout waiting for unregister device: %@", device.nickname);
        result = [NSError errorWithDomain:@"Virtuoso" code:kVDE_InternalError userInfo:@{NSLocalizedDescriptionKey: @"Timeout waiting for device delete to complete."}];
    }
    
    return result;
}



- (IBAction)deleteClicked:(UIButton *)sender
{
    sender.enabled = FALSE;  // block reentry
    if (self.exampleAsset)
    {
        [self.exampleAsset deleteAssetOnComplete:^
         {
             self.statusLabel.text = @"";
             self.statusProgressBar.progress = 0;
             // Give next smartdownload time to get started
             dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 1.5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                 [self loadEngineData];
             });
             
         }];
    }
}

- (IBAction)playBtnClicked:(UIButton *)sender
{
    if (self.exampleAsset)
    {
        [self playAsset:self.exampleAsset];
    }
}

- (IBAction)downloadBtnClicked:(UIButton *)sender
{
    sender.enabled = NO; // block reentrancy while we create the asset
    
    if ( kVDE_ReachableViaWiFi != [VirtuosoDownloadEngine.instance currentNetworkStatus] ) {
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Network Unreachable" message:@"Demo requires WiFi" preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            sender.enabled = YES;
        }]];
         [self presentViewController:alert animated:TRUE completion:nil];
         return;
    }
    
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        // Important: AssetID should be unique across your video catalog
        NSString* myAssetID = @"SEASON-1-EPISODE-1";
        
        // Create asset configuration object
        VirtuosoAssetConfig* config =
            [[VirtuosoAssetConfig alloc] initWithURL:@"http://virtuoso-demo-content.s3.amazonaws.com/bbb/season1/ep1/index.m3u8"
                                             assetID:myAssetID
                                         description:@"The Right Kite"
                                                type:kVDE_AssetTypeHLS];
        
        if (!config)
        {
            NSLog(@"create config failed");
            sender.enabled = YES;
            return;
        }

        [VirtuosoAsset assetWithConfig:config];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            sender.enabled = YES;
            [self refreshView];
        });
        
    });
}

//
// MARK: Internal implementation
//

- (void)displayAsset:(VirtuosoAsset*)asset
{
    self.assetName.text = [NSString stringWithFormat:@"Episode: %@", asset.description];
    
    double fractionComplete = asset.fractionComplete;
    kVDE_DownloadStatusType status = asset.status;
    if( status != kVDE_DownloadComplete && status != kVDE_DownloadProcessing )
    {
        NSString* statusMessage = (asset.status == kVDE_DownloadInitializing) ? NSLocalizedString(@"Initializing: ", @"") : @"";
        NSString* errorsNessage = (asset.downloadRetryCount > 0)? [NSString stringWithFormat:@" (Errors: %i)",asset.downloadRetryCount] : @"";
        self.statusLabel.text = [NSString stringWithFormat:@"%@%0.02f%% (%qi MB)%@", statusMessage, fractionComplete*100.0, asset.currentSize/1024/1024, errorsNessage];
        self.statusProgressBar.progress = fractionComplete;
        [self.downloadBtn setEnabled:NO];
    }
    else if( status == kVDE_DownloadComplete )
    {
        self.statusLabel.text = [NSString stringWithFormat:@"%0.02f%% (%qi MB)",fractionComplete*100.0,asset.currentSize/1024/1024];
        self.statusProgressBar.progress = fractionComplete;
    }
}

- (void)playAsset:(VirtuosoAsset*)asset {
    [VirtuosoAsset isPlayable:asset callback:^(Boolean playable) {
        if( playable )
        {
            if( asset.type != kVDE_AssetTypeHSS )
            {
                UIViewController* player = [DemoPlayerViewController new];
                self.exampleAsset = asset;
                
                [asset playUsingPlaybackType:kVDE_AssetPlaybackTypeLocal
                                   andPlayer:(id<VirtuosoPlayer>)player
                                   onSuccess:^ {
                                       // Present the player
                                       [self presentViewController:player animated:YES completion:nil];
                                   }
                                      onFail:^ {
                                          self.exampleAsset = nil;
                                          self.error = nil;
                                          NSLog(@"Can't play video!");
                                      }];
            }
        }
    }];
}

// ------------------------------------------------------------------------------------------------------------
// IMPORTANT:
// Loads data from the Download Engine.
// ------------------------------------------------------------------------------------------------------------
- (void)loadEngineData {

    self.error = nil;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        //
        // Invoked on background thread to prevent blocking UI updates
        //
        if( [[VirtuosoDownloadEngine instance]started] ) {
            NSArray* downloadComplete = [VirtuosoAsset completedAssetsWithAvailabilityFilter:NO];
            NSArray* downloadPending = [VirtuosoAsset pendingAssetsWithAvailabilityFilter:NO];
            
            dispatch_async(dispatch_get_main_queue(), ^{
                //
                // Switch back to MainThread to update UI
                //
                if ( nil != downloadComplete && downloadComplete.count > 0 )
                {
                    self.exampleAsset = (VirtuosoAsset *)downloadComplete.firstObject;
                }
                else if (nil != downloadPending && downloadPending.count > 0) {
                    self.exampleAsset = (VirtuosoAsset *)downloadPending.firstObject;
                }
                else
                {
                    self.exampleAsset = nil;
                }
                [self refreshView];
            });
        }
        else {
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                [self loadEngineData];
            });
        }

    });
}


-(void)refreshView
{
    if (nil == self.exampleAsset) {
        
        self.statusLabel.text = @"Ready to download";
        self.statusProgressBar.progress = 0;
        
        [self setEnabledAppearance:self.downloadBtn enabled:YES];
        [self setEnabledAppearance:self.playBtn enabled:NO];
        [self setEnabledAppearance:self.deleteBtn enabled:(nil != self.error ? TRUE : FALSE)];
        
        return;
    }
    
    Boolean downloadInProcess = (self.exampleAsset.fractionComplete < 1) ? true : false;
    self.statusLabel.text = (downloadInProcess ? @"Downloading..." : @"Ready to play");
    
    [self setEnabledAppearance:self.deleteBtn enabled:YES];
    [self setEnabledAppearance:self.downloadBtn enabled:NO];
    self.statusProgressBar.progress = self.exampleAsset.fractionComplete;

    [VirtuosoAsset isPlayable:self.exampleAsset callback:^(Boolean playable) {
        [self setEnabledAppearance:self.playBtn enabled:playable];
    }];
}


-(void)setEnabledAppearance:(UIControl*)control enabled:(Boolean)enabled
{
    [control setEnabled:enabled];
    [control setBackgroundColor:(enabled ? [UIColor blackColor] : [UIColor lightGrayColor])];
}

//
// MARK: VirtuosoDownloadEngineNotificationsDelegate - required methods ONLY
//

// ------------------------------------------------------------------------------------------------------------
//  Called whenever the Engine starts downloading a VirtuosoAsset object.
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineDidStartDownloadingAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self displayAsset:asset];
}

// ------------------------------------------------------------------------------------------------------------
// Called whenever the Engine reports progress for a VirtuosoAsset object
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineProgressUpdatedForAsset:(VirtuosoAsset * _Nonnull)asset
{
    [VirtuosoAsset isPlayable:asset callback:^(Boolean playable) {
        if (playable)
        {
            self.exampleAsset = asset;
            [self refreshView];
        }
        [self displayAsset:asset];
    }];
}

// ------------------------------------------------------------------------------------------------------------
// Called when an asset is being processed after background transfer
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineProgressUpdatedProcessingForAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self displayAsset:asset];
}

// ------------------------------------------------------------------------------------------------------------
// Called whenever the Engine reports a VirtuosoAsset as complete
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineDidFinishDownloadingAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self displayAsset:asset];
    [self refreshView];
    [self loadEngineData];
}

- (void)downloadEngineDidEncounterErrorForAsset:(VirtuosoAsset *)asset error:(NSError *)error task:(NSURLSessionTask *)task data:(NSData *)data statusCode:(NSNumber *)statusCode
{
    self.error = error;
    [self displayAsset:asset];
    [self refreshView];
    [self loadEngineData];
}

- (void)downloadEngineInternalQueueUpdate {
    [self loadEngineData];
}

// ------------------------------------------------------------------------------------------------------------
// Called when Engine start is complete
// ------------------------------------------------------------------------------------------------------------
-(void)downloadEngineStartupComplete:(Boolean)succeeded {
    [self loadEngineData];
}

// Device was unregistered
- (void)backplaneDidUnregisterDeviceWithStatus:(Boolean)success error:(NSError * _Nullable)error
{
    self.detectedDeviceUnregister = true;
}

// Device is about to start remote kill
- (void)backplaneStartingRemoteKill
{
    
}

// Device was remote killed happened
- (void)backplaneRemoteKill
{
    
}

// Device sync completed
- (void)backplaneSyncCompleteWithStatus:(Boolean)status error:(NSError * _Nullable)error
{
    
}


@end

File: Tutorials/Swift/Example11/Example11/ViewController.swift
//
//  ViewController.swift
//  Example11
//
//  Created by dev on 1/31/20.
//  Copyright © 2020 Penthera. All rights reserved.
//

import UIKit

class ViewController: UIViewController, VirtuosoDownloadEngineNotificationsDelegate, VirtuosoBackplaneNotificationsDelegate {
    var detectedDeviceUnregister = false
    
    
    // <-- change these to your settings in production
    let backplaneUrl = "replace_with_your_backplane_url"
    let publicKey = "replace_with_your_public_key"
    let privateKey = "replace_with_your_private_key"
    
    //
    // MARK: Instance data
    //
    var exampleAsset: VirtuosoAsset?
    var downloadEngineNotifications: VirtuosoDownloadEngineNotificationManager!
    var backplaneNotifications: VirtuosoBackplaneNotificationsManager!
    var hasDownloaded = false
    var error: Error?
    
    @IBOutlet weak var statusLabel: UILabel!
    @IBOutlet weak var playBtn: UIButton!
    @IBOutlet weak var downloadBtn: UIButton!
    @IBOutlet weak var deleteBtn: UIButton!
    @IBOutlet weak var statusProgressBar: UIProgressView!
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        // download engine update listener
        downloadEngineNotifications = VirtuosoDownloadEngineNotificationManager.init(delegate: self)
        backplaneNotifications = VirtuosoBackplaneNotificationsManager.init(delegate: self)
        
        self.statusLabel.text = "Starting Engine..."

        //
        // Enable the Engine
        VirtuosoDownloadEngine.instance().enabled = true;
        
        // Backplane permissions require a unique user-id for the full range of captabilities support to work
        // Production code that needs this will need a unique customer ID.
        // For demonstation purposes only, we use the device name
        let userName = UIDevice.current.name
        
        //
        // Create the engine confuration
        guard let config = VirtuosoEngineConfig(user: userName,
                                                backplaneUrl: self.backplaneUrl,
                                                publicKey: self.publicKey,
                                                privateKey: self.privateKey)
        else
        {
            DispatchQueue.main.async {
                let alert = UIAlertController(title: "Setup Required", message: "Please contact support@penthera.com to setup the backplaneUrl, pulicKey, and privateKey", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                    exit(0)
                }))
                self.present(alert, animated: true, completion: nil)
            }
            return
        }

        //
        // Start the Engine
        // This method will execute async, the callback will happen on the main-thread.
        VirtuosoDownloadEngine.instance().startup(config) { (status) in
            if status != .vde_EngineStartupSuccess {
                let alert = UIAlertController(title: "Startup Failed", message: "Startup completed with status: \(status)", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                }))
                self.present(alert, animated: true, completion: nil)
            }
        }
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        super.viewWillAppear(animated)
    }
    
    func displayAsset(asset: VirtuosoAsset)
    {
        let fractionComplete = asset.fractionComplete
        let status = asset.status
        var errorAdd = ""
        if( asset.downloadRetryCount > 0 ) {
            errorAdd = String(format: " (Errors: %i)", asset.downloadRetryCount)
        }
        if status != .vde_DownloadComplete && status != .vde_DownloadProcessing {
            let optionalStatus = (asset.status == .vde_DownloadInitializing) ? NSLocalizedString("Initializing: ", comment: "") : ""
            statusLabel.text = String(format: "%@ %@%0.02f%% (%qi MB)%@", asset, optionalStatus, fractionComplete*100.0, asset.currentSize/1024/1024, errorAdd)
            statusProgressBar.progress = Float(fractionComplete)
            downloadBtn.isEnabled = false
        }
        else if status == .vde_DownloadComplete {
            statusLabel.text = String(format: "%@ %0.02f%% (%qi MB)", asset, fractionComplete*100.0, asset.currentSize/1024/1024)
            statusProgressBar.progress = Float(fractionComplete)
        }
    }
    
    private func loadEngineData()
    {
        self.error = nil

        DispatchQueue.global().async {
            if VirtuosoDownloadEngine.instance().started {
                let downloadComplete = VirtuosoAsset.completedAssets(withAvailabilityFilter: false)
                let downloadPending = VirtuosoAsset.pendingAssets(withAvailabilityFilter: false)
                DispatchQueue.main.async {
                    if downloadComplete.count > 0 {
                        self.exampleAsset = downloadComplete.first as? VirtuosoAsset
                    }
                    else if downloadPending.count > 0 {
                        self.exampleAsset = downloadPending.first as? VirtuosoAsset
                    }
                    else {
                        self.exampleAsset = nil
                    }
                    self.refreshView()
                }
            }
            else {
                DispatchQueue.main.async {
                    self.loadEngineData()
                }
            }
        }
    }
    
    private func refreshView()
    {
        guard let asset = self.exampleAsset else {
            self.statusLabel.text = "Ready to download"
            self.downloadBtn.backgroundColor = UIColor.black
            self.playBtn.backgroundColor = UIColor.lightGray
            self.downloadBtn.isEnabled = true
            self.playBtn.isEnabled = false
            
            if nil != self.error {
                deleteBtn.isEnabled = true
                deleteBtn.backgroundColor = UIColor.black
            } else {
                deleteBtn.isEnabled = false
                deleteBtn.backgroundColor = UIColor.lightGray
            }

            self.statusProgressBar.progress = 0
            return
        }
        
        if asset.fractionComplete < 1.0 {
            statusLabel.text = "Downloading..."
        } else {
            statusLabel.text = "Ready to play"
        }
        
        deleteBtn.isEnabled = true
        deleteBtn.backgroundColor = UIColor.black
        downloadBtn.isEnabled = false
        downloadBtn.backgroundColor = UIColor.lightGray
        
        statusProgressBar.progress = Float(asset.fractionComplete)

        VirtuosoAsset.isPlayable(asset) { (playable) in
            if playable {
                self.playBtn.isEnabled = true
                self.playBtn.backgroundColor = UIColor.black
            } else {
                self.playBtn.isEnabled = false
                self.playBtn.backgroundColor = UIColor.lightGray
            }
        }
    }
    
    //
    // Handles unregistering devices
    @IBAction func unregisterDevicesClicked(_ sender: Any)
    {
        DispatchQueue.global().async {
            let error = self.unregisterAllDevices()
            
            var maximumWait:Double = 30;
            let sleepDuration:Double = 1.0;
            while (!self.detectedDeviceUnregister && maximumWait > 0)
            {
                Thread.sleep(forTimeInterval: sleepDuration)
                maximumWait -= sleepDuration;
            }

            DispatchQueue.main.async {
                let alertView: UIAlertController?
                if nil != error
                {
                    //
                    // Encountered an error while attempting to unregister,
                    // most likely cause is a Network issue.
                    alertView = UIAlertController(title: "Unregister Failed", message: "Error encountered while unregistering. error: \(String(describing: error))", preferredStyle: .alert)
                    alertView?.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                    }))
                }
                else if (maximumWait <= 0) {
                    alertView = UIAlertController(title: "Unregister Complete With Issues", message: "All devices are unregistered but never detected device unregister notification. App will now exit.", preferredStyle: .alert)
                    alertView?.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                        exit(0);
                    }))
                }
                else
                {
                    //
                    // Devices un-registered
                    // Penthera SDK must be initialzed again before it can be used.
                    // See: VirtuosoDownloadEngine.instance().startup
                    alertView = UIAlertController(title: "Unregister Complete", message: "All devices are unregistered. App will now exit.", preferredStyle: .alert)
                    alertView?.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                        exit(0)
                    }))
                }
                guard let alert = alertView else { return }
                self.present(alert, animated: true, completion: nil)
            }
        }
    }

    func unregisterAllDevices() -> NSError? {
        // Private function to handle unregistering device
        // and waiting for that to complete.
        func unregisterAndWait(device : VirtuosoDevice) ->NSError?
        {
            let semaphore = DispatchSemaphore(value:0)
            var result: NSError? = nil
            DispatchQueue.global().async {
                device.unregister { (status, error) in
                    
                    result = error as NSError?
                    if (status)
                    {
                        print("unregister device \(device.nickname) succeeded.")
                    }
                    else
                    {
                        print("unregister device \(device.nickname) failed.")
                    }
                    semaphore.signal()
                }
            }
            semaphore.wait()
            
            return result
        }

        
        // Important:
        // Once you have unregistered devices, Penthera is unavailable until
        // you invoke VirtuosoDownloadEngine.instance().startup again.
        
        let devices = VirtuosoDownloadEngine.instance().devices;
        
        var currentDevice: VirtuosoDevice? // last device to unregister is the current device
        var error: NSError? = nil
        for anyObject in devices
        {
            if (anyObject is VirtuosoDevice)
            {
                let device = anyObject as! VirtuosoDevice
                if device.isThisDevice
                {
                    currentDevice = device
                }
                else
                {
                    error = unregisterAndWait(device: device)
                    if error != nil {
                        return error
                    }
                }
            }
        }
        
        guard let current = currentDevice else {
            return nil
        }
        
        // unregister the current device
        return unregisterAndWait(device: current)

    }
    
    @IBAction func deleteBtnClicked(_ sender: UIButton)
    {
        sender.isEnabled = false   // Block reentry
        
        guard let asset = self.exampleAsset else { return }
        asset.delete {
            self.statusLabel.text = ""
            self.statusProgressBar.progress = 0
            // Give next smartdownload time to get started
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.5, execute: {
                self.loadEngineData()
            })
        }
    }
    
    @IBAction func playBtnClicked(_ sender: UIButton)
    {
        guard let asset = self.exampleAsset else {
            print("asset not playable yet")
            return
        }
        
        VirtuosoAsset.isPlayable(asset) { (playable) in
            if !playable { return }

            switch(asset.type) {
            case .vde_AssetTypeHLS, .vde_AssetTypeDASH, .vde_AssetTypeNonSegmented:
                let demoPlayer = DemoPlayerViewController()
                
                asset.play(using: .vde_AssetPlaybackTypeLocal, andPlayer: demoPlayer as VirtuosoPlayer, onSuccess: {
                    self.present(demoPlayer, animated: true, completion: nil)
                    
                }) {
                    self.exampleAsset = nil
                    self.error = nil
                    print("Video is unplayble.")
                }
                break
                
            case .vde_AssetTypeHSS:
                print("Playback of HSS supported in another Tutorial")
                break
                
            @unknown default:
                break;
            }
        }
    }
    
    @IBAction func downloadBtnClicked(_ sender: UIButton)
    {
        sender.isEnabled = false // block reentrancy while we create the asset
        
        if .vde_ReachableViaWiFi != VirtuosoDownloadEngine.instance().currentNetworkStatus() {
            let alertView = UIAlertController(title: "Network Unreachable", message: "Demo requires WiFi", preferredStyle: .alert)
            alertView.addAction(UIAlertAction(title: "Ok", style: .cancel, handler: { (action) in
                sender.isEnabled = true
            }))
            self.present(alertView, animated: true, completion: nil)
            return
        }
        
        // Create the Asset on a background thread
        DispatchQueue.global(qos: .background).async {
            // Create asset configuration object
            guard let config = VirtuosoAssetConfig(url: "http://virtuoso-demo-content.s3.amazonaws.com/bbb/season1/ep1/index.m3u8",
                                                   assetID: "SEASON-1-EPISODE-1",
                                                   description: "The Right Kite",
                                                   type: kVDE_AssetType.vde_AssetTypeHLS) else {
                                                    print("create config failed")
                                                    sender.isEnabled = true
                                                    return
            }
            
            // Create the Playlist
            // The playlist contains three assets, which will download in sequence as previous episode
            // is played and then deleted.
            //
            // PlaylistConfig is used to control how smart-download works.
            let playlistConfig = VirtuosoPlaylistConfig(name: "TEST_PLAYLIST")
            
            // Create array of Playlist
            var playlists : Array<VirtuosoPlaylist> = Array()
            
            // Add a Playlist to the array
            playlists.append(VirtuosoPlaylist(config: playlistConfig,
                                              assets: ["SEASON-1-EPISODE-1", "SEASON-1-EPISODE-2", "SEASON-1-EPISODE-3"]))
            
            // When you create an Asset, you can also create playlists that might be assocated with the Asset.
            config.playlists = playlists
            
            // For demo purposes we clear the previous playlist.
            // This causes 'TEST_PLAYLIST' to be removed completely, resetting the demo.
            // Ordinarily you would NOT do this as doing so will reset the smart-downloading cycle,
            // restarting at the beginning.
            VirtuosoPlaylistManager.instance().clear("TEST_PLAYLIST")
            
            // Create asset and commence downloading.
            let _ = VirtuosoAsset.init(config: config)
            
            //
            // Alternatively, you can create a Playlist directly using the following:
            // VirtuosoPlaylistManager.instance().create(withItems: playlists)
            
            //
            // Playlists can be appended to as well
            // VirtuosoPlaylistManager.instance().appendItems(playlists)

            DispatchQueue.main.async {
                sender.isEnabled = true
                self.hasDownloaded = true
                self.refreshView()
            }
        }
        
    }

    //
    // MARK: VirtuosoDownloadEngineNotificationsDelegate - required methods ONLY
    //
    
    // --------------------------------------------------------------
    //  Called whenever the Engine starts downloading a VirtuosoAsset object.
    // --------------------------------------------------------------
    func downloadEngineDidStartDownloadingAsset(_ asset: VirtuosoAsset)
    {
        displayAsset(asset: asset)
    }
    
    // --------------------------------------------------------------
    // Called whenever the Engine reports progress for a VirtuosoAsset object
    // --------------------------------------------------------------
    func downloadEngineProgressUpdated(for asset: VirtuosoAsset)
    {
        DispatchQueue.global(qos: .background).async {
            VirtuosoAsset.isPlayable(asset) { (playable) in
                if playable {
                    self.exampleAsset = asset;
                    self.refreshView()
                }
                self.displayAsset(asset: asset)
            }
        }
    }
    
    // --------------------------------------------------------------
    // Called when an asset is being processed after background transfer
    // --------------------------------------------------------------
    func downloadEngineProgressUpdatedProcessing(for asset: VirtuosoAsset)
    {
        displayAsset(asset: asset)
    }
    
    // --------------------------------------------------------------
    // Called whenever the Engine reports a VirtuosoAsset as complete
    // --------------------------------------------------------------
    func downloadEngineDidFinishDownloadingAsset(_ asset: VirtuosoAsset)
    {
        // Download is complete
        displayAsset(asset: asset)
        refreshView()
        loadEngineData()
    }
    
    // --------------------------------------------------------------
    // Called whenever an Engine encounters error downloading asset
    // --------------------------------------------------------------
    func downloadEngineDidEncounterError(for asset: VirtuosoAsset, error: Error?, task: URLSessionTask?, data: Data?, statusCode: NSNumber?) {
        self.error = error
        displayAsset(asset: asset)
        refreshView()
        loadEngineData()
    }
    
    // --------------------------------------------------------------
    // Called whenever an asset is added to the Engine
    // --------------------------------------------------------------
    func downloadEngineInternalQueueUpdate() {
        loadEngineData()
    }
    
    // --------------------------------------------------------------
    // Called whenever Engine start completes
    // --------------------------------------------------------------
    func downloadEngineStartupComplete(_ succeeded: Bool) {
        loadEngineData()
    }
    
    
    //
    // MARK: VirtuosoBackplaneNotificationsDelegate - required methods ONLY
    //
    
    func backplaneSyncComplete(withStatus status: Bool, error: Error?) {
        
    }

    func backplaneRemoteKill() {
        
    }
    
    func backplaneStartingRemoteKill() {
        
    }
    
    func backplaneDidUnregisterDevice(withStatus success: Bool, error: Error?) {
        self.detectedDeviceUnregister = true
    }
    

}

